src/CMakeLists.txt

```diff
@@ -35,18 +18,148 @@ add_subdirectory(../lib/parms ${PROJECT_BINARY_DIR}/parms)
+  oce_ale_pressure_bv.F90
+  oce_ale_tracer.F90
+  oce_ale_vel_rhs.F90
+  oce_dyn.F90
+  oce_fer_gm.F90
+  oce_ice_init_state.F90
+  oce_local.F90
+  oce_mesh.F90
+  oce_mo_conv.F90
+  oce_modules.F90
+  oce_muscl_adv.F90
+  oce_setup_step.F90
+  oce_shortwave_pene.F90
+  oce_spp.F90
+  oce_tracer_mod.F90
+  oce_vel_rhs_vinv.F90
+  psolve.c
+  psolve_feom.c
+  toy_channel_soufflet.F90
+  write_step_info.F90
+)
+target_compile_definitions(fesom PRIVATE
+  PARMS
+  -DMETIS_VERSION=5 -DPART_WEIGHTED -DMETISRANDOMSEED=35243
+  $<$<BOOL:${DISABLE_MULTITHREADING}>:DISABLE_MULTITHREADING>
 
-add_library(${PROJECT_NAME}_C ${sources_C})
-target_compile_definitions(${PROJECT_NAME}_C PRIVATE PARMS USE_MPI REAL=double DBL HAS_BLAS FORTRAN_UNDERSCORE VOID_POINTER_SIZE_8 SGI LINUX UNDER_ MPI2)
-target_link_libraries(${PROJECT_NAME}_C parms) #metis
+  $<$<COMPILE_LANGUAGE:C>:
+    USE_MPI REAL=double DBL HAS_BLAS FORTRAN_UNDERSCORE VOID_POINTER_SIZE_8 SGI LINUX UNDER_ MPI2
+   >
+)
 
-# create our binary (set its name to name of this project)
-add_executable(${PROJECT_NAME} ${sources_Fortran})
-target_compile_definitions(${PROJECT_NAME} PRIVATE PARMS -DMETIS_VERSION=5 -DPART_WEIGHTED -DMETISRANDOMSEED=35243)
-if(${DISABLE_MULTITHREADING})
-   target_compile_definitions(${PROJECT_NAME} PRIVATE DISABLE_MULTITHREADING)
-endif()
+# CMAKE_Fortran_COMPILER_ID will also work if a wrapper is being used (e.g. mpif90 wraps ifort -> compiler id is Intel)
+# https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html
+# https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_ID.html#variable:CMAKE_%3CLANG%3E_COMPILER_ID
+target_compile_options(fesom PRIVATE
+  $<$<BOOL:${FESOM_COUPLED}>:__oasis>
+  $<$<BOOL:${OIFS_COUPLED}>:__oifs>
+  $<$<BOOL:${VERBOSE}>:VERBOSE>
+
+  # or use $<COMPILE_LANG_AND_ID:language,compiler_ids>
+  $<$<COMPILE_LANG_AND_ID:Fortran,Intel>:
+    -r8 -i4 -fp-model precise -no-prec-div -no-prec-sqrt -fast-transcendentals -xHost -ip -init=zero -no-wrap-margin
+   >
+
+  $<$<COMPILE_LANG_AND_ID:Fortran,GNU>:
+    -O3 -finit-local-zero  -finline-functions -march=native -fimplicit-none  -fdefault-real-8 -ffree-line-length-none
+    $<$<VERSION_GREATER_EQUAL:${CMAKE_Fortran_COMPILER_VERSION},10>:
+      -fallow-argument-mismatch # gfortran v10 is strict about erroneous API calls: "Rank mismatch between actual argument at (1) and actual argument at (2) (scalar and rank-1)"
+     >
+   >
+
+  $<$<COMPILE_LANG_AND_ID:Fortran,Cray>:
+    -c -emf -hbyteswapio -hflex_mp=conservative -hfp1 -hadd_paren -Ounroll0 -hipa0 -r am -s real64
+   >
+
+  $<$<COMPILE_LANG_AND_ID:C,PGI>:
+    -fast -fastsse -O3 -Minline -Munroll
+   >
+  $<$<COMPILE_LANG_AND_ID:Fortran,PGI>:
+   -fast -fastsse -O3 -Mallocatable=95 -Minline -Munroll -Mr8 -pgf90libs
+    $<$<BOOL:${FESOM_OPENACC}>:
+      -acc -ta=tesla:cc35
+      -Minfo=accel
+     >
+   >
+  )
+
+target_link_options(fesom PRIVATE
+  $<$<Fortran_COMPILER_ID:PGI>:
+    -acc -ta=tesla:cc35
+  >
+)
```

Comment:
maybe #515 could integrate this to simplify the option definitions: we see them all at first glance!

Reply:
@dguibert apologies again for not being proactive. it  would be great if you can please check if what you suggest is relevant in refactoring and production_DE branches if you find anything can you please make a PR to those branches respectively. I wonder merging to current master which will soon be replaced by refactoring is not worth your effort and time.